### PR TITLE
Create Config::Cookie Class

### PR DESCRIPTION
**Pull request by mostafakhudair**
While we are taking a step forward to make config folder separated as much as possible, Cookie class should have it's own config file

### CODE REVIEW

**Comment by MGatner** [comment]
This is a good change. I would like to make a release this weekend, which would solidify all the Cookie changes since 4.1 - let's be sure to get this finished so it can be included.

**Comment by mostafakhudair** [comment]
@MGatner I think next release should be done without any of new Cookie Class, because of cookie class is bloated with alot usless methods and need to be refactored, also to avoid BC

**Comment by MGatner** [comment]
@mostafakhudair I thought there had been significant improvements to our Cookie handling since the last release, work that you and @paulbalandan both did. What do you think should be held back? Is it something that can easily be reverted and re-applied for the following release?

Paul, what are your thoughts?

**Comment by paulbalandan** [comment]
Bloated? What aspects of the `Cookie` are bloated? Cookie is just implementing its contract interfaces' methods, which are actually getters and cloners of a web cookie's attributes. Plus, it has bonus methods to help parse a `Cookie-Header` string into a functional `Cookie` object, and some validation methods to make sure the passed attributes are valid against the multitude of cookie specifications. Are these useless to you? The bloat, I would say IMO, is the `ArrayAccess` implementation and some `?:` logic, but that is a compromise to retain backward compatibility to the time cookie handling was just using plain arrays.

If you can pinpoint the places that needs refactor, we'd consider excluding cookie in the release and perhaps I'd do the refactor myself. Otherwise, aside from the config reassignment, the cookie class is stable as-is.

File: system/Cookie/Cookie.php [hunk 1]
```diff
@@ -629,7 +631,7 @@ public function withSameSite(string $sameSite)
 	 */
 	public function offsetExists($offset)
 	{
-		if (in_array($offset, ['expire', 'httponly', 'samesite'], true))
+		if (in_array($offset, ['expires', 'httponly', 'samesite'], true))
```
Comment (paulbalandan) [inline system/Cookie/Cookie.php:634]: This is a BC break.

File: system/Cookie/Cookie.php [hunk 2]
```diff
@@ -653,19 +655,19 @@ public function offsetGet($offset)
 			throw new InvalidArgumentException(sprintf('Undefined offset "%s".', $offset));
 		}
 
-		if ($offset === 'expire')
+		if ($offset === 'expires')
```
Comment (paulbalandan) [inline system/Cookie/Cookie.php:658]: BC break.

**Review by paulbalandan — CHANGES_REQUESTED** [review]

File: app/Config/App.php [hunk 3]
```diff
@@ -310,40 +319,11 @@ class App extends BaseConfig
...
-	 *
-	 * If this is set to `true`, cookie names should be compliant of RFC 2616's
-	 * list of allowed characters.
-	 *
-	 * @var boolean
-	 *
-	 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#attributes
-	 * @see https://tools.ietf.org/html/rfc2616#section-2.2
-	 */
-	public $cookieRaw = false;
```
Comment (samsonasik) [inline app/Config/App.php:332]: this is BC break, existing user no longer can't use `(new Config\App())->cookieRaw;`

Comment (mostafakhudair) [reply app/Config/App.php:332]: (new Config\App())->cookieRaw; wasn't released yet,

Comment (samsonasik) [reply app/Config/App.php:332]: oh, ok 👍

**Comment by samsonasik** [comment]
@mostafakhudair please incorporate @paulbalandan review

File: app/Config/Cookie.php [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	 * - ''
+	 *
+	 * Alternatively, you can use the constant names:
+	 * - `Cookie::SAMESITE_NONE`
+	 * - `Cookie::SAMESITE_LAX`
+	 * - `Cookie::SAMESITE_STRICT`
+	 *
+	 * Defaults to `Lax` for compatibility with modern browsers. Setting `''`
+	 * (empty string) means default SameSite attribute set by browsers (`Lax`)
+	 * will be set on cookies. If set to `None`, `$cookieSecure` must also be set.
```
Comment (paulbalandan) [inline app/Config/Cookie.php:96]: $cookieSecure change to $secure

File: system/Cookie/Cookie.php [hunk 5]
```diff
@@ -118,28 +119,29 @@ class Cookie implements ArrayAccess, CloneableCookieInterface
...
-				'raw'      => $config->cookieRaw ?? false,
-				'samesite' => $config->cookieSameSite ?? self::SAMESITE_LAX,
+				'prefix'   => $config->prefix,
+				'expires'  => $config->expires ?? 0,
+				'path'     => $config->path,
+				'domain'   => $config->domain,
+				'secure'   => $config->secure,
+				'httponly' => $config->httponly,
+				'samesite' => $config->samesite ?? self::SAMESITE_LAX,
+				'raw'      => $config->raw ?? false,
```
Comment (paulbalandan) [inline system/Cookie/Cookie.php:144]: No need for the fall back `??` values since you are using the new config.

File: system/Cookie/Cookie.php [hunk 6]
```diff
@@ -660,12 +662,12 @@ public function offsetGet($offset)
...
 		{
-			return $this->httpOnly;
+			return $this->httponly;
 		}
 
 		if ($offset === 'samesite')
 		{
-			return $this->sameSite;
+			return $this->samesite;
 		}
```
Comment (paulbalandan) [inline system/Cookie/Cookie.php:671]: Since you made all property names as lowercase, then there's no  need for these 2 guards here. Also in offsetExists method, no need to check for the two. Only check for `expire`.

File: system/Session/Session.php [hunk 7]
```diff
@@ -178,22 +179,27 @@ public function __construct(SessionHandlerInterface $driver, App $config)
...
-			'domain'   => $config->cookieDomain ?? '',
-			'path'     => $config->cookiePath ?? '/',
-			'secure'   => $config->cookieSecure ?? false,
+			'path'     => $this->cookiePath,
+			'domain'   => $this->cookieDomain,
+			'secure'   => $this->cookieSecure,
 			'httponly' => true, // for security
-			'samesite' => $config->cookieSameSite ?? Cookie::SAMESITE_LAX,
+			'samesite' => $this->cookieSameSite,
+			'raw'      => $cookie->raw ?? false,
```
Comment (paulbalandan) [inline system/Session/Session.php:202]: Do not add dependence to the deprecated cookie properties. It is deprecated so that it will no longer be used in active code and can be safely removed in the future but still preserving for those users who depended on such. By deprecating but still actively depending on those properties would defeat the purpose of the deprecation.

File: system/Test/Mock/MockCookieConfig.php [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\Test\Mock;
+
+use Config\Cookie;
+
+class MockCookieConfig extends Cookie
```
Comment (paulbalandan) [inline system/Test/Mock/MockCookieConfig.php:16]: Do not add useless mocks. Mocks can be easily made natively with PHPUnit's mocking system.

File: tests/system/Cookie/CookieTest.php [hunk 9]
```diff
@@ -173,24 +174,40 @@ public function testSetCookieHeaderCreation(string $header, array $changed): voi
...
-		];
+				  'test=value; Secure; HttpOnly; SameSite=Lax',
+				  [
+					  'name'     => 'test',
+					  'value'    => 'value',
+					  'secure'   => true,
+					  'httponly' => true,
+					  'samesite' => 'Lax',
+				  ],
+			  ];
```
Comment (paulbalandan) [inline tests/system/Cookie/CookieTest.php:210]: Revert the changes here. our codesniffer implementation cannot still handle gracefully these nested arrays but can become unsightly.

File: tests/system/Security/SecurityTest.php [hunk 10]
```diff
@@ -209,12 +209,12 @@ public function testSendingCookiesFalse(): void
 		$request = $this->createMock(IncomingRequest::class);
 		$request->method('isSecure')->willReturn(false);
 
-		$config = new MockAppConfig();
+		$config = config('Cookie');
```
Comment (paulbalandan) [inline tests/system/Security/SecurityTest.php:212]: I'm not sure if we have proper cleanup of shared instances in the tests, but please refrain from using shared instances. Instead, use the `injectMock` facility of `Factories` with a subsequent `reset` to clean up.

The thing is using `config(...)` calls in the tests and altering the state will persist and affect the other tests depending on another `config(..)` calls. As much as possible, we want to make sure each and every test receives a fresh, unadulterated instance of a config it needs.

**Review by paulbalandan — CHANGES_REQUESTED** [review]

File: app/Config/Cookie.php (hunk 4 above)
Comment (mostafakhudair) [reply app/Config/Cookie.php:96]: shoud we force set secure to true when developer uses None a samesite attribute?

File: tests/system/Cookie/CookieTest.php (hunk 9 above)
Comment (mostafakhudair) [reply tests/system/Cookie/CookieTest.php:210]: Actually I didn't change this, maybe my IDE changed it, reverted

File: app/Config/Cookie.php (hunk 4 above)
Comment (paulbalandan) [reply app/Config/Cookie.php]: That is in the cookie spec. None samesite should have the Secure attribute. Otherwise, the cookie is rejected.

Comment (mostafakhudair) [reply app/Config/Cookie.php]: what is this happend?

if (strtolower($samesite) === self::SAMESITE_NONE && ! $secure)
{
	//throw CookieException::forInvalidSameSiteNone();
	self::setDefaults(['secure' => true];
}

Comment (paulbalandan) [reply app/Config/Cookie.php]: That would be condoning the developer's mistake. What sets apart CI4 from CI3 is that we throw exceptions for coding mistakes rather than the framework correcting them and have the dev with no idea that they have a mistake.

File: system/Session/Session.php [hunk 11]
```diff
@@ -181,19 +182,24 @@ public function __construct(SessionHandlerInterface $driver, App $config)
...
-			'domain'   => $config->cookieDomain ?? '',
-			'path'     => $config->cookiePath ?? '/',
-			'secure'   => $config->cookieSecure ?? false,
+			'path'     => $config->path,
+			'domain'   => $config->domain,
+			'secure'   => $config->secure,
 			'httponly' => true, // for security
-			'samesite' => $config->cookieSameSite ?? Cookie::SAMESITE_LAX,
+			'samesite' => $config->samesite,
+			'raw'      => $config->raw ?? false,
```
Comment (paulbalandan) [inline system/Session/Session.php:202]:
```suggestion
			'raw'      => $config->raw,
```
```suggestion
			'raw'      => $config->raw ?? false,
```

File: tests/system/Cookie/CookieTest.php (hunk 9 above)
Comment (mostafakhudair) [reply tests/system/Cookie/CookieTest.php:210]: Reverted more than once, but when I commit and pull, it auto reverse my changes

Comment (paulbalandan) [reply tests/system/Cookie/CookieTest.php:210]: Are you using VS Code?

Comment (paulbalandan) [reply tests/system/Cookie/CookieTest.php:210]: You can use the Git Staged Amend No Verify option. Then Push Force

Comment (paulbalandan) [reply tests/system/Cookie/CookieTest.php:210]: if not. directly in terminal
```
git commit --amend --no-verify
git push --force-with-lease
```

Comment (mostafakhudair) [reply tests/system/Cookie/CookieTest.php]: Thank you

**Comment by mostafakhudair** [comment]
What is wrong with tests?

**Comment by samsonasik** [comment]
That seems race condition issue in SQLSRV action. It seems unrelated.

**Comment by MGatner** [comment]
Have tests been rerun since #4539?

**Comment by mostafakhudair** [comment]
Yes, twice

**Review by paulbalandan — APPROVED** [review]

**Comment by paulbalandan** [comment]
Thank you, @mostafakhudair

**Comment by MGatner** [comment]
🥳
